### PR TITLE
Page content UI

### LINKED ISSUE #52: Implement page content UI and evaluate Netlify
[Wireframe](https://lucid.app/lucidchart/6be33dd8-f7db-47f8-8af8-338255f23c3c/edit?shared=true&page=hHG8TGWUOjR0#?folder_id=home&browser=icon)

This will be the initial implementation of the page where authors will add/edit site content.

Here's an example of the initial mapping we have in mind from the site config to the database record:

```yml
  - label: "Page"
    name: "page"
    fields:
      - {label: "Title", name: "title", widget: "string"}
      - {label: "Body", name: "body", widget: "markdown"}
      - {label: "Field 1", name: "field-1", widget: "string"}
      - {label: "Field 2", name: "field-2", widget: "string"}
```
```python
WebsiteContent.title = "My Title"
WebsiteContent.markdown = "## My page\n\n### Subsection 1\n\n..."
WebsiteContent.metadata = {
  "field-1": "Value 1",
  "field-2": "Value 2",
}
```

In other words, a config field named `title` maps to the `WebsiteContent.title` field, a config field named `body` maps to the `WebsiteContent.markdown` field, and all other fields are saved in `metadata` as JSON as `{[field_name]: [value]}`. The `populate_local_starters` command can be used to load in local site configs.

We will be evaluating [Netlify widgets](https://www.netlifycms.org/docs/widgets/#default-widgets) ([Github](https://github.com/netlify/netlify-cms/tree/master/packages)) for this project. For this issue, we just need widgets for simple text and for markdown. We're still evaluating CKEditor (#39) for use as our WYSIWYG editor. If we decide that CKEditor suits our needs, we will replace the Netlify markdown editor with that. It would help to also start evaluating the other widgets they offer.

**Acceptance criteria:**
- This page is rendered when the "Content" submenu items are clicked (see #40)
- Routes added for adding a new piece of content and editing an existing piece of content
- Content editing form has widgets that match the `fields` element of the site config
  - The ordering of these fields should match the config, and they can be rendered vertically in a single column
- Saving the form creates/updates a WebsiteContent record with the values in the form fields
- If page content listing UI is done before this, the "Add" buttons and content type list items link to this page
- Evaluate Netlify text widget and markdown widget for use in OCWS

**Not needed:**
- Widget types other than text/string and markdown
- Field validation

### LINKED ISSUE #56: Add page content listing UI
**Depends on site detail UI: https://github.com/mitodl/ocw-studio/issues/40**
Can be worked on in parallel with content authoring UI: https://github.com/mitodl/ocw-studio/issues/52 

Wireframe [1](https://lucid.app/lucidchart/6be33dd8-f7db-47f8-8af8-338255f23c3c/edit?shared=true&page=8AATMRnQLEi0#?folder_id=home&browser=icon), [2](https://lucid.app/lucidchart/6be33dd8-f7db-47f8-8af8-338255f23c3c/edit?shared=true&page=erront7o~TLZ#?folder_id=home&browser=icon)

**Acceptance criteria:**
- When content type is selected in the site detail UI ("Pages", "Resources"), objects of that type are shown in the main content window
  - If content authoring UI is already done, these items should link to the content authoring page where the content can be edited. Otherwise the items don’t need to link to anything
- "Add" button added in main content pane
  - If content authoring UI is already done, this button should navigate to the content authoring page. Otherwise this button can be a no-op

**Not needed:**
- Dot menus
- File types as shown in the “Resources” wireframe

### PR DESCRIPTION
**Pull request by noisecapella**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [ ] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #52 
Fixes #56 

#### What's this PR do?
Adds front-end pages for adding and editing site content. 

#### How should this be manually tested?
 - Listing content: go to `/sites/site-goes-here/` and click on Page or Resource in the sidebar. You should see a list of content with titles and an edit button on the side.
 - Add new content: click on "Add Page" or "Add Resource" on the upper right. Write in some simple content. If you're adding a page you should see the markdown editor. For resource you should just see a description textbox.
   - Leave title blank and click "Save". You should see a error message about a missing title.
   - Write in a title and click "Save". You should be returned to the content listing and you should see your new content at the bottom of the list. (If you refresh this will show up at the top of the list since the sort is by most recently modified).
 - Click "Edit" next to the item you just created. A drawer should open from the right and whatever you added before should appear there. 
   - You should be able to dismiss the drawer by clicking outside the drawer or clicking on the 'x' on the upper right.
   - Clear the title text field and click "Save". You should see a warning that the title is missing.
   - Make a change to the title and to the other content, and click "Save". The drawer should close and you should see that the content you modified has a new title in the list. If you click "Edit" next to it again the drawer should reopen and all of your changes that you saved should be there.

#### Other context
 - I left out mobile support since there are no wireframes yet
 - I'm not sure how permissions factor into the UI here so I left that out too

#### Screenshots
Screenshot from 2021-03-03 08-51-35

Screenshot from 2021-03-02 16-44-29

Screenshot from 2021-03-03 08-50-52

### CODE REVIEW

**Comment by noisecapella** [comment]
@gsidebo @Ferdi I got `netlify-cms-widget-string` working. See the text input with the blue border:
Screenshot from 2021-03-03 11-23-19

However, I'm having trouble with `netlify-cms-widget-markdown`. There is a [format](https://www.netlifycms.org/docs/custom-widgets/#registerwidget) which widget packages need to conform to, but nothing is documented about how  widgets can be used independent of netlify-cms. Unlike `netlify-cms-widget-string` which just wraps an `input`, there are a lot of bits and pieces to `netlify-cms-widget-markdown` so I think it would be more effort than it's worth to figure out, especially since ckeditor is working for this purpose

**Comment by Ferdi** [comment]
@noisecapella looks great.  ok, feel free to ignore netlifycms widgets.  Hopefully we can come up with something ourselves that is reusable and maintainable .

**Comment by gsidebo** [comment]
I'm thinking we should have the same UI for adding and editing rather than a form on a separate page for adding and a right side panel for editing. @Ferdi I know this UI is just wireframe, but I think it would be best to settle on a single UI for both options. Personally I think we should go with a separate page for both since page content could be quite long. Any thoughts

**Comment by Ferdi** [comment]
@gsidebo I understand the concern.  It feels a bit inconsistent.   Side bar is probably good for uploading a file and adding some metadata quick. But maybe not ideal for authoring long pages. 

We should be able to come up with something that accommodate both use cases, but I think we should go ahead and merge this PR.

**Comment by gsidebo** [comment]
@noisecapella Can you rebase?

**Comment by aliceriot** [comment]
unicode support seems to be working!


Screenshot from 2021-03-09 13-49-23

File: static/js/lib/urls.ts [hunk 1]
```diff
@@ -22,3 +22,10 @@ export const siteApiCollaboratorsDetailUrl = (
   name: string,
   username: string
 ): string => `${siteApiCollaboratorsUrl(name)}${username}/`
+export const siteAddContentUrl = (name: string, configName: string): string =>
+  `/sites/${name}/${configName}/add/`
```
Comment (aliceriot) [inline static/js/lib/urls.ts:26]: is maybe `contentType` possibly a more accurate name than `configName` here?

Comment (aliceriot) [reply static/js/lib/urls.ts:26]: just wondering based on how I'm seeing it used

File: static/js/components/forms/SiteAddForm.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      {field.label}
+    </label>
+    <Field
+      as={componentFromWidget(field)}
+      name={field.name}
+      className="form-control"
+    />
+    <ErrorMessage name={field.name} component="div" />
+  </div>
+)
```
Comment (aliceriot) [inline static/js/components/forms/SiteAddForm.tsx:28]: You have some very similar code in `SiteEditForm.tsx` below - could be helpful to pull this out as a separate `SiteContentField` component that just takes a single prop of type `ConfigField`

File: static/js/components/SiteContentListing.tsx [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+import { websiteContentListingRequest } from "../query-configs/websites"
+import {
+  getWebsiteDetailCursor,
+  getWebsiteContentListingCursor
+} from "../selectors/websites"
+
+import { ConfigItem, WebsiteContentListItem } from "../types/websites"
 
 interface MatchParams {
   configname: string
```
Comment (aliceriot) [inline static/js/components/SiteContentListing.tsx:18]: I'd suggest that we perhaps rename `configname` here too, it just doesn't seem to capture actually what this value represents to me. What do you think?

File: static/js/selectors/websites.ts [hunk 4]
```diff
@@ -26,3 +27,15 @@ export const getWebsiteCollaboratorDetailCursor = createSelector(
       find(propEq("username", username), collaborators[name])
     )
 )
+
+export const getWebsiteContentListingCursor = createSelector(
+  (state: ReduxState) => state.entities?.websiteContentListing ?? {},
+  listing => (name: string, type: string) =>
+    // not memoized since there are two arguments
```
Comment (aliceriot) [inline static/js/selectors/websites.ts:34]: it's possible to still memoize with two arguments, you just need to supply a custom function to determine how to memoize (i.e. how to convert arguments into a key for the cache the `memoize` function uses).

there's an example here in Open: https://github.com/mitodl/open-discussions/blob/master/static/js/lib/queries/learning_resources.js#L187

**Review by aliceriot — CHANGES_REQUESTED** [review]
a few questions and comments after reading through

File: static/js/lib/urls.ts (hunk 1 above)
Comment (noisecapella) [reply static/js/lib/urls.ts:26]: Yes that makes sense to me

File: static/js/components/SiteContentListing.tsx (hunk 3 above)
Comment (noisecapella) [reply static/js/components/SiteContentListing.tsx]: Sure, I decided just to do a blanket rename of that so everything is the same

**Comment by noisecapella** [comment]
@aliceriot should be ready for review again

**Review by gumaerc — APPROVED** [review]
:+1: LGTM

**Pull request merged**